File: src/autonomous-trust/conformance/harness/python/runner.py
# ******************
#  Copyright 2026 Sean M. Brennan and contributors
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
# ******************

"""Python harness for the AT conformance corpus.

Owns dispatch from `Case` -> per-protocol adapter, collects `CaseResult`
records, and emits a result JSON in the corpus's `results/` directory.
"""

from __future__ import annotations

import json
import subprocess
import time
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from ..common.scenario_loader import Case, discover

from .adapters.agreement import AgreementAdapter
from .adapters.bootstrap import BootstrapAdapter
from .adapters.identity import IdentityAdapter
from .adapters.negotiation import NegotiationAdapter
from .adapters.network import NetworkAdapter
from .adapters.reputation import ReputationAdapter


CORPUS_ROOT = Path(__file__).resolve().parents[2]


@dataclass
class CaseResult:
    case_id: str
    status: str  # pass | fail | skip
    duration_ms: int = 0
    reason_class: str | None = None
    rejected_at_step: int | None = None
    detail: str | None = None

    def to_dict(self) -> dict[str, Any]:
        d = {k: v for k, v in asdict(self).items() if v is not None}
        return d


@dataclass
class Report:
    implementation: str = 'python'
    implementation_version: str = field(default_factory=lambda: _git_sha())
    schema_version: str = '1'
    started_at: str = field(default_factory=lambda: datetime.now(timezone.utc).isoformat())
    cases: list[CaseResult] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return {
            'implementation': self.implementation,
            'implementation_version': self.implementation_version,
            'schema_version': self.schema_version,
            'started_at': self.started_at,
            'cases': [c.to_dict() for c in self.cases],
        }


def _git_sha() -> str:
    try:
        out = subprocess.check_output(
            ['git', 'rev-parse', '--short=12', 'HEAD'],
            cwd=str(CORPUS_ROOT),
            stderr=subprocess.DEVNULL,
        )
        return out.decode('ascii').strip()
    except (subprocess.CalledProcessError, FileNotFoundError):
        return 'unknown'


# Adapter registry. The protocol field of a case picks the adapter; each
# adapter handles every kind it knows about (raising NotImplementedError
# for kinds it does not yet support).
def _adapters() -> dict[str, Any]:
    return {
        'network': NetworkAdapter(corpus_root=CORPUS_ROOT),
        'identity': IdentityAdapter(corpus_root=CORPUS_ROOT),
        'agreement': AgreementAdapter(corpus_root=CORPUS_ROOT),
        'negotiation': NegotiationAdapter(corpus_root=CORPUS_ROOT),
        'reputation': ReputationAdapter(corpus_root=CORPUS_ROOT),
        'bootstrap': BootstrapAdapter(corpus_root=CORPUS_ROOT),
    }


def run_case(case: Case, adapters: dict[str, Any]) -> CaseResult:
    adapter = adapters.get(case.protocol)
    if adapter is None:
        return CaseResult(case_id=case.case_id, status='skip',
                          detail=f'no adapter for protocol {case.protocol!r}')
    method = {
        'wire_vector': 'run_wire_vector',
        'crypto_vector': 'run_crypto_vector',
        'scenario': 'run_scenario',
        'negative': 'run_negative',
        'agreement_vector': 'run_agreement_vector',
    }.get(case.kind)
    if method is None or not hasattr(adapter, method):
        return CaseResult(case_id=case.case_id, status='skip',
                          detail=f'adapter {type(adapter).__name__} does not handle kind {case.kind!r}')
    t0 = time.monotonic()
    try:
        getattr(adapter, method)(case)
        ms = int((time.monotonic() - t0) * 1000)
        return CaseResult(case_id=case.case_id, status='pass', duration_ms=ms)
    except NotImplementedError as exc:
        return CaseResult(case_id=case.case_id, status='skip', detail=str(exc))
    except AssertionError as exc:
        ms = int((time.monotonic() - t0) * 1000)
        return CaseResult(case_id=case.case_id, status='fail',
                          duration_ms=ms, detail=str(exc) or 'assertion failed')
    except Exception as exc:  # noqa: BLE001 — runner needs broad capture
        ms = int((time.monotonic() - t0) * 1000)
        return CaseResult(case_id=case.case_id, status='fail',
                          duration_ms=ms, reason_class=type(exc).__name__,
                          detail=f'{type(exc).__name__}: {exc}')


def run_all() -> Report:
    """Discover and execute every case under CORPUS_ROOT. Returns the report."""
    cases = discover(CORPUS_ROOT)
    adapters = _adapters()
    report = Report()
    for case in cases:
        report.cases.append(run_case(case, adapters))
    return report


def write_report(report: Report) -> Path:
    out_dir = CORPUS_ROOT / 'results'
    out_dir.mkdir(parents=True, exist_ok=True)
    ts = datetime.now(timezone.utc).strftime('%Y%m%dT%H%M%SZ')
    out_path = out_dir / f'python-{ts}.json'
    out_path.write_text(json.dumps(report.to_dict(), indent=2) + '\n', encoding='utf-8')
    return out_path


if __name__ == '__main__':
    report = run_all()
    out = write_report(report)
    failed = [c for c in report.cases if c.status == 'fail']
    print(f'wrote {out}: {len(report.cases)} cases, {len(failed)} failed')
    for c in failed:
        print(f'  FAIL {c.case_id}: {c.detail}')
    raise SystemExit(1 if failed else 0)

File: src/c/conformance/runner.c
/********************
 *  Copyright 2026 TekFive, Inc., Sean M. Brennan, and contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *   You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 *******************/

/** @file Conformance runner — C harness entry point.
 *
 *  Usage: conformance_runner <corpus_json_root> <results_out_path>
 *
 *  - corpus_json_root: directory containing index.json + scenarios/ + vectors/,
 *    produced by `python -m tools.corpus_to_json`.
 *  - results_out_path: where to write the per-run results JSON. Format
 *    matches the Python harness so the diff tool can compare apples to
 *    apples.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <jansson.h>
#include <sodium.h>

#include "scenario_loader.h"
#include "case_result.h"
#include "adapters/network.h"
#include "adapters/identity.h"
#include "adapters/agreement.h"
#include "adapters/negotiation.h"
#include "adapters/reputation.h"
#include "adapters/bootstrap.h"
#include "adapters/physics.h"
#include "adapters/calibration.h"
#include "adapters/certificate.h"

/* Adapters that handle kind:negative need the JSON corpus root to resolve
 * `based_on` references; one runner invocation processes one root, so a
 * process-global is the simplest plumbing. */
static const char *g_corpus_json_root = NULL;
const char *at_runner_corpus_json_root(void) { return g_corpus_json_root; }

/* Adapter dispatch keyed on protocol. Every adapter knows what kinds it
 * handles and emits skip for anything else. */
static void dispatch(const at_case_t *c, at_case_result_t *out) {
    if (strcmp(c->protocol, "network") == 0) {
        at_network_run(c, out);
    } else if (strcmp(c->protocol, "identity") == 0) {
        at_identity_run(c, out);
    } else if (strcmp(c->protocol, "agreement") == 0) {
        at_agreement_run(c, out);
    } else if (strcmp(c->protocol, "negotiation") == 0) {
        at_negotiation_run(c, out);
    } else if (strcmp(c->protocol, "reputation") == 0) {
        at_reputation_run(c, out);
    } else if (strcmp(c->protocol, "bootstrap") == 0) {
        at_bootstrap_run(c, out);
    } else if (strcmp(c->protocol, "physics") == 0) {
        at_physics_run(c, out);
    } else if (strcmp(c->protocol, "certificate") == 0) {
        at_certificate_run(c, out);
    } else if (strcmp(c->protocol, "calibration") == 0) {
        at_calibration_run(c, out);
    } else {
        char detail[160];
        snprintf(detail, sizeof(detail),
                 "no C adapter for protocol %s (kind=%s)",
                 c->protocol, c->kind);
        at_case_result_set_skip(out, detail);
    }
}

static const char *iso8601_now(char *buf, size_t n) {
    time_t t = time(NULL);
    struct tm tm;
    gmtime_r(&t, &tm);
    strftime(buf, n, "%Y-%m-%dT%H:%M:%SZ", &tm);
    return buf;
}

static int write_results(const char *path,
                         const at_case_result_t *results, size_t n) {
    json_t *report = json_object();
    json_object_set_new(report, "implementation", json_string("c"));
    /* Implementation version: leave as "unknown" for now; CI can stamp the
     * git sha externally if needed. */
    json_object_set_new(report, "implementation_version", json_string("unknown"));
    json_object_set_new(report, "schema_version", json_string("1"));
    char ts[32];
    iso8601_now(ts, sizeof(ts));
    json_object_set_new(report, "started_at", json_string(ts));

    json_t *cases = json_array();
    for (size_t i = 0; i < n; i++) {
        json_array_append_new(cases, at_case_result_to_json(&results[i]));
    }
    json_object_set_new(report, "cases", cases);

    int rc = json_dump_file(report, path, JSON_INDENT(2));
    json_decref(report);
    return rc;
}

int main(int argc, char **argv) {
    if (argc < 3) {
        fprintf(stderr, "usage: %s <corpus_json_root> <results_out_path>\n", argv[0]);
        return 2;
    }
    if (sodium_init() < 0) {
        fprintf(stderr, "sodium_init failed\n");
        return 1;
    }

    const char *corpus = argv[1];
    const char *results_path = argv[2];
    g_corpus_json_root = corpus;

    at_case_t *cases = NULL;
    size_t n = 0;
    if (at_load_corpus(corpus, &cases, &n) != 0) {
        fprintf(stderr, "failed to load corpus from %s\n", corpus);
        return 1;
    }

    at_case_result_t *results = calloc(n, sizeof(at_case_result_t));
    if (results == NULL) {
        fprintf(stderr, "oom\n");
        return 1;
    }

    size_t pass = 0, fail = 0, skip = 0;
    for (size_t i = 0; i < n; i++) {
        at_case_result_init(&results[i], cases[i].case_id);
        dispatch(&cases[i], &results[i]);
        switch (results[i].status) {
            case AT_PASS: pass++; break;
            case AT_FAIL: fail++; break;
            case AT_SKIP: skip++; break;
        }
    }

    if (write_results(results_path, results, n) != 0) {
        fprintf(stderr, "failed to write results to %s\n", results_path);
    } else {
        printf("wrote %s: %zu cases, %zu pass, %zu fail, %zu skip\n",
               results_path, n, pass, fail, skip);
    }

    /* Print failed cases inline for quick visibility. */
    for (size_t i = 0; i < n; i++) {
        if (results[i].status == AT_FAIL) {
            fprintf(stderr, "  FAIL %s: %s\n",
                    results[i].case_id,
                    results[i].detail ? results[i].detail : "(no detail)");
        }
    }

    int exit_code = (fail > 0) ? 1 : 0;

    for (size_t i = 0; i < n; i++) {
        at_case_result_free(&results[i]);
        at_case_free(&cases[i]);
    }
    free(results);
    free(cases);
    return exit_code;
}
